src/lib.rs

```diff
@@ -30,7 +42,143 @@ pub(crate) const SEMANTIC_DE_FN: &str = "de_fn";
 pub(crate) const START_VERSION: &str = "start";
 pub(crate) const END_VERSION: &str = "end";
 
-/// Implements the derive proc macro.
+/// Generates serialization an deserialization code as an implementation of
+/// the `Versionize` trait.
+/// 
+/// Different code paths are generated for each version of the structure or 
+/// enum. There is no limit enforced on the maximum number of structure versions.
+/// 
+/// ### Struct and enum requirements
+/// - all members or enum variants need to implement the `Versionize` trait
+/// - no generics are being used (this is currenly a limitation)
+/// 
+/// ## Annotations
+/// 
+/// To facilitate version tolerant serialization "history metadata" is attached
+/// to the structure or enum. This is done by using the `version` attribute in 
+/// their definition. In the below example a new field is added to the 
+/// structure starting with version 2: `#[version(start = 2)]`.
+/// 
+/// ```rust
+/// #[derive(Versionize)]
+/// struct Test {
+///     a: u32,
+///     #[version(start = 2)]
+///     b: u8,
+/// }
+/// ```
+/// 
+/// Multiple version annotations can be defined for a field, like for example:
+/// `#[version(start = 2, end = 3)]`. Field was added in structure version 2 and 
+/// removed in version 3. The generated code will attempt to (de)serialize this field
+/// only for version 2 of the structure.
+/// 
+/// ### Supported field attributes and usage
+/// 
+/// The `version` attribute accepts multiple key/value pairs to be specified in order 
+/// to support versioning, semantic serialization and default values for fields. All 
+/// of these are optional and a default behaviour is provided in their absence.
+/// 
+/// #### default_fn 
+/// 
+/// Provides an initialization value for a field when deserializing from an older 
+/// structure version which does not contain this field. If not specified the `Default`
+/// trait isused to initialize the field.
+/// ```rust
+/// #[derive(Versionize)]
+/// struct Test {
+///     a: u32,
+///     #[version(start = 2, default_fn = "default_b")]
+///     b: u8,
+/// }
+/// 
+/// impl TestStruct {
+///     fn default_b(_source_version: u16) -> u8 {
+///         12u8
+///     }
+/// }
+/// ```
+/// 
+/// The function name needs to be specified as a string and it's prototype must
+/// take an u16 source version parameter and return a value of the same type as
+/// as the field.
+/// 
+/// #### start/end
+/// 
+/// Defines the field version lifetime. Fields can be added by specifing the start
+/// version of the structure when first defining them and can be later on removed 
+/// from serialization logic by adding and end version. 
+/// 
+/// For example: `#[version(start = 2, end = 4)]`. The field would be present in 
+/// the structure v2 and v3, but starting with v4 it would no longer be serialized
+/// or deserialized. 
+/// 
+/// Once a field is removed, it can never be added again at future version.
```

Comment:
Just a curiosity, if you want to add back a removed field, it should be added with a different name in the best case?

Reply:
Yes, that's one way of doing it.